### DISCUSSION TITLE
delegate* need to add managed hasthis support!

### DISCUSSION THREAD

**Discussion by Dogwei** [dotnet/csharplang#4067] [Language Ideas]
![微信截图_20201026121152](https://user-images.githubusercontent.com/14845492/97146618-1f780700-17a3-11eb-877b-b8536e089df6.png)

As shown in the picture, the managed calling convention cannot call the get method of Program.Guid correctly.
Need to use CallingConventions.HasThis to call correctly, but delegate* does not provide a calling convention similar to managed_hasthis.
CallingConventions.HasThis is different from CallingConvention.ThisCall in that the former is managed and the latter is unmanaged.

**Comment by 333fred** [comment]
Note: it's pretty easy to work around this by making a static method to wrap the instance method.

```cs
using System;
using System.Runtime.CompilerServices;

unsafe
{
    var obj = new Program();
    delegate*<Program, Guid> ptr = &Program.ProgramGuid;
    Console.WriteLine(ptr(obj));
}

public class Program
{
    public static Guid ProgramGuid(Program program) => program.Guid;
    
    public Guid Guid
    {
        get
        {
            
            Console.WriteLine("this address : " + Unsafe.As<Program, IntPtr>(ref Unsafe.AsRef(this)));
        
            var value = Guid.NewGuid();
            
            Console.WriteLine("value : " + value);
            
            return value;
        }
    }
}
```
```
this address : 140226118852704
value : a6487b48-6a28-4bf2-8ae3-65b36cd7232f
a6487b48-6a28-4bf2-8ae3-65b36cd7232f
```

**Reply by Dogwei** [reply]
However, as a class library developer, we cannot require users to write indirect static methods.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md

> Restrictions of this feature:
>
> - Only applies to methods marked as static.

**Reply by NMSAzulX** [reply]
@333fred  Will there be more updates on lightweight delegate pointers later?

**Reply by 333fred** [reply]
@NMSAzulX we have no plans currently.

**Reply by hamarb123** [reply]
I'd also like for `HasThis` and `ExplicitThis` to be added.
